[581] Defer students

avinhurry opened the pull request:
### Context

The ability for providers to defer students who have been registered but not yet completed their programme. Not including DTTP interaction.

### Changes proposed in this pull request

- 2 new controllers to handle the deferral details and confirmation journeys

- All the views and components necessary to capture and display information

### Guidance to review

- Pull down the branch and fire up the server
- Create a trainee and complete all the sections
- Navigate to http://localhost:3000/trainees/{{trainee-id?}}/edit
- Click the defer button
- Have a play around, check that all the buttons work as expected 
- Check the correct validation error messages are displayed if an error is made
- Check that it matches the prototype

avinhurry commented:
![](https://github.trello.services/images/mini-trello-icon.png) [[M] Defer students](https://trello.com/c/GLbpojvn/581-m-defer-students)

edwardhorsford commented:
Had a quick play.

Spotted two issues:

Defer link incorrectly styled (same issue with withdrawal)
![Screenshot 2020-12-22 at 11 07 01](https://user-images.githubusercontent.com/2204224/102882198-ee841d80-4445-11eb-81a7-051ff22fdad7.png)

If you submit the date page with no options selected, the radio should be shown in error - but instead the conditional field is in error.
![Screenshot 2020-12-22 at 11 06 15](https://user-images.githubusercontent.com/2204224/102882250-0065c080-4446-11eb-867f-4673723779f7.png)

edwardhorsford commented:
Is the deferral happening before the confirmation page? if so we should suppress the cancel link until such time as it actually cancels.

I don't see the state changing to 'deferred' after doing the flow - should it be?

AbigailMcP commented on app/components/trainees/record_actions/view.html.erb:
```diff
@@ -1,2 +1,7 @@
+<%= govuk_button_link_to "Record training outcome", edit_trainee_outcome_details_outcome_date_path(trainee),
+                         { class: "govuk-!-margin-bottom-0" }  %>
+
+<%= govuk_link_to "Defer", trainee_deferral_path(@trainee), class: "defer"  %>
```
The styling issue here that @edwardhorsford  spotted should be fixed if you merge in master now.

avinhurry replied:
This is styled correctly now

AbigailMcP commented on app/views/trainees/confirm_deferrals/show.html.erb:
```diff
@@ -0,0 +1,30 @@
+    <%= render Trainees::DeferralDetails::View.new(@trainee) %>
+
+    <%= form_with url: trainee_confirm_deferral_path(@trainee), method: :put, local: true do |f| %>
+      <%= hidden_field_tag :trainee_id, @trainee.id %>
```
What's this hidden field tag for? You shouldn't need this I don't think.

avinhurry replied:
Thanks, removed.

AbigailMcP commented on app/views/trainees/confirm_deferrals/show.html.erb:
```diff
@@ -0,0 +1,30 @@
+  </div>
+</div>
+
+<p class="govuk-body"><%= govuk_link_to("Cancel and return to record", edit_trainee_path(@trainee)) %></p>
```
Remove this for now as per Ed's comment.

avinhurry replied:
I've removed the cancel link from `app/views/trainees/deferrals/show.html.erb`, I think that's the one you meant?

AbigailMcP replied:
I mean the one on the confirmation page, not on the show page. Once you get to the confirmation page, the deferral has already been saved to the database. So until we fix that the 'Cancel' button on the confirmation page is misleading.

avinhurry replied:
oh cool I see, sorry, i've removed the one from the confirmation page

AbigailMcP commented on app/views/trainees/deferrals/show.html.erb:
```diff
@@ -0,0 +1,32 @@
+
+<%= content_for(:breadcrumbs) do %>
+  <%= render GovukComponent::BackLink.new(
+    text: 'Back',
```
In John's PR he added a new translation for 'Back' links. For consistency, it would be good to use that here (and in your other 'Back' link on the confirm page.

avinhurry replied:
Thanks, fixed this

AbigailMcP commented on app/views/trainees/deferrals/show.html.erb:
```diff
@@ -0,0 +1,32 @@
+      </span>
+
+      <%= f.govuk_radio_buttons_fieldset(:defer_date_string, legend: { text: "When did the trainee defer?", tag: "h1", size: "l" }, classes: "outcome-date") do %>
+        <%= f.govuk_radio_button :defer_date_string, :today, label: { text: "Today" }, link_errors: true %>
```
John also added common translations for "Today", "Yesterday", "Specific date" and "On what date". Please use them here.

Bonus points for also updating the outcome date form whilst you're there!

avinhurry replied:
Thanks. Have updated this, will address outcome date in a separate PR

AbigailMcP commented on db/migrate/20201221173705_add_defer_date_to_trainees.rb:
```diff
@@ -0,0 +1,7 @@
+
+class AddDeferDateToTrainees < ActiveRecord::Migration[6.0]
+  def change
+    add_column :trainees, :defer_date, :date
```
We're storing the `withdraw_date` as a Datetime. Do we also want to keep the time information on deferrals? I might 'defer to' @jhenderson on this one.

jhenderson replied:
I might have made a mistake using `datetime`. I say go with `date` and I'll fix `withdraw_date` in the UUID PR.

AbigailMcP commented on spec/components/trainees/confirmation/deferral_details/view.rb:
```diff
@@ -0,0 +1,21 @@
+# frozen_string_literal: true
```
I don't think you meant to have this file here! Can just be deleted.

avinhurry replied:
Thanks

AbigailMcP commented on spec/components/trainees/confirmation/deferral_details/view_preview.rb:
```diff
@@ -0,0 +1,29 @@
+        render(Trainees::DeferralDetails::View.new(mock_trainee("ABC-1234-XYZ")))
+      end
+
+      def with_no_trainee_id
```
I think you might have copied this over from another preview file - your view doesn't render the trainee_id anywhere, so we don't need a preview for this case.

Just a default route will suffice. Pass in a trainee with the deferral date, similar to `spec/components/trainees/confirmation/withdrawal_details/view_preview.rb`.

avinhurry replied:
Thank you, changed this

AbigailMcP commented on app/views/trainees/deferrals/show.html.erb:
```diff
@@ -0,0 +1,28 @@
+        <%= f.govuk_radio_button :defer_date_string, :today, label: { text: t("views.forms.common.today") }, link_errors: true %>
+        <%= f.govuk_radio_button :defer_date_string, :yesterday, label: { text: t("views.forms.common.yesterday") } %>
+        <%= f.govuk_radio_button :defer_date_string, :other, label: { text: t("views.forms.common.specific_date") } do %>
+          <%= f.govuk_date_field :defer_date, legend: { text: 'On what date?', size: 's' }, hint: { text: "For example, 3 12 2020" } %>
```
There's a translation for 'On what date?' too.

AbigailMcP commented on app/views/trainees/outcome_dates/edit.html.erb:
```diff
@@ -17,9 +17,9 @@
+        <%= f.govuk_radio_button :outcome_date_string, :today, label: { text: t("views.forms.common.today") }, link_errors: true %>
+        <%= f.govuk_radio_button :outcome_date_string  , :yesterday, label: { text: t("views.forms.common.yesterday") } %>
+        <%= f.govuk_radio_button :outcome_date_string, :other, label: { text: t("views.forms.common.specific_date") } do %>
           <%= f.govuk_date_field :outcome_date, legend: { text: 'On what date?', size: 's' }, hint: { text: "For example, 3 12 2020" } %>
```
Ditto re this translation.

AbigailMcP commented on app/views/trainees/outcome_dates/edit.html.erb:
```diff
@@ -17,9 +17,9 @@
-        <%= f.govuk_radio_button :outcome_date_string, :yesterday, label: { text: "Yesterday" } %>
-        <%= f.govuk_radio_button :outcome_date_string, :other, label: { text: "On another day" } do %>
+        <%= f.govuk_radio_button :outcome_date_string, :today, label: { text: t("views.forms.common.today") }, link_errors: true %>
+        <%= f.govuk_radio_button :outcome_date_string  , :yesterday, label: { text: t("views.forms.common.yesterday") } %>
```
```suggestion
        <%= f.govuk_radio_button :outcome_date_string, :yesterday, label: { text: t("views.forms.common.yesterday") } %>
```

avinhurry replied:
looks like the linter has fixed this

avinhurry commented:
Had a conversation offline with @edwardhorsford. 

- Fixed the styling for the `Defer` button
- Removed the cancel button from the confirm page
- Agreed to fix the error styling as part of a separate PR as this now applies to all three places with this field.
